File: dwarf/targets/cardano-node/src/DecodeKeepAlive.hs
{-# LANGUAGE OverloadedStrings #-}
-- | cardano-node-mini-protocol-decode-keep-alive shim.
-- Decodes a KeepAlive mini-protocol wire-format message
-- (MsgKeepAlive cookie, MsgKeepAliveResponse cookie, or MsgDone).
-- Wire format from ouroboros-network KeepAlive codec_v2:
--   MsgKeepAlive          = [0, cookie:u16]
--   MsgKeepAliveResponse  = [1, cookie:u16]
--   MsgDone               = [2]
module Main (main) where

import qualified Codec.CBOR.Decoding             as CBOR
import qualified Codec.CBOR.Read                 as CBOR
import qualified Data.ByteString.Lazy            as BSL
import           System.Exit                     (ExitCode (..), exitSuccess, exitWith)

decoder :: CBOR.Decoder s ()
decoder = do
  len <- CBOR.decodeListLen
  key <- CBOR.decodeWord
  case (len, key) of
    (2, 0) -> CBOR.decodeWord16 >> pure ()
    (2, 1) -> CBOR.decodeWord16 >> pure ()
    (1, 2) -> pure ()
    _      -> fail ("unexpected (len, key) = (" ++ show len ++ ", " ++ show key ++ ")")

main :: IO ()
main = do
  bytes <- BSL.getContents
  case CBOR.deserialiseFromBytes decoder bytes of
    Left err          -> do
      putStrLn ("ERR " ++ show err)
      exitWith (ExitFailure 1)
    Right (rest, ()) | BSL.null rest -> do
      putStrLn "OK"
      exitSuccess
    Right _ -> do
      putStrLn "ERR trailing bytes after message"
      exitWith (ExitFailure 1)

File: dwarf/targets/amaru/src/bin/decode_keep_alive.rs
// amaru-mini-protocol-decode-keep-alive shim.
// Decodes a KeepAlive mini-protocol wire-format message:
//   MsgKeepAlive          = [0, cookie:u16]
//   MsgKeepAliveResponse  = [1, cookie:u16]
//   MsgDone               = [2]
// using minicbor (already a transitive dep through amaru-kernel/pallas).

use std::io::{self, Read};
use std::process::ExitCode;

fn decode(data: &[u8]) -> Result<(), String> {
    let mut d = minicbor::Decoder::new(data);
    let len = d.array().map_err(|e| format!("array: {e}"))?
        .ok_or_else(|| "indefinite-length array not allowed".to_string())?;
    let key: u32 = d.u32().map_err(|e| format!("key: {e}"))?;
    match (len, key) {
        (2, 0) | (2, 1) => {
            let _cookie: u16 = d.u16().map_err(|e| format!("cookie: {e}"))?;
        }
        (1, 2) => {}
        _ => return Err(format!("unexpected (len, key) = ({len}, {key})")),
    }
    if d.position() != data.len() {
        return Err(format!("trailing bytes at position {}", d.position()));
    }
    Ok(())
}

fn main() -> ExitCode {
    let mut buf = Vec::new();
    if let Err(e) = io::stdin().read_to_end(&mut buf) {
        println!("ERR stdin read: {e}");
        return ExitCode::from(1);
    }
    match decode(&buf) {
        Ok(()) => {
            println!("OK");
            ExitCode::from(0)
        }
        Err(msg) => {
            println!("ERR {msg}");
            ExitCode::from(1)
        }
    }
}
